### PR TITLE
LIE compression

### PR DESCRIPTION
**Pull request by endJunction**
Prevents negative apertures

Review can be done commit-wise.

TODO:

- [x] Include provided documentation into web/content
- [x] Add the two new projects (negative aperture and displacement controlled) to ctests.

### CODE REVIEW

**Comment by nagelt** [comment]
@endJunction Documentation here ...
[doku_lie_comp.pdf](https://github.com/ufz/ogs/files/1334650/doku_lie_comp.pdf)

**Comment by nagelt** [comment]
@endJunction You can use our test. Note that the cutoff should be negative (see last Eq. in docu). Rationale is that our test would lead to negative apertures but doesn't with this implementation.

**Comment by nagelt** [comment]
@renchao-lu could you have a look at this PR, please?

**Comment by renchao-lu** [comment]
@nagelt @endJunction It's my pleasure. Also, it's a good exercise for me to get familiar with this branch.

File: MaterialLib/FractureModels/FractureModelBase.h [hunk 1]
```diff
@@ -66,6 +66,7 @@ class FractureModelBase
      *
      * @param t           current time
      * @param x           current position in space
+     * @param aperture0   initial fracture's aperture
```
Comment (renchao-lu) [inline MaterialLib/FractureModels/FractureModelBase.h]: I would name this variable as aperture_bm0 instead which physically means the initial mechanical aperture at a certain node or initial mean mechanical aperture at a representative element, thereby leaving some naming space for other relevant variables, for instance, aperture_bh0 standing for initial hydraulic aperture.

Comment (renchao-lu) [reply MaterialLib/FractureModels/FractureModelBase.h]: @endJunction @nagelt The variable "aperture0" introduced here keeps consistent with the pre-existing definition in the header file "FractureProperty.h". Therefore, just leave my previous immature thinking out.

File: MaterialLib/FractureModels/LogPenalty.h [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+/// A penalty function for negative aperture supression used as a multiplier to
+/// the normal fracture stiffness.
+///
+/// The derivative is continuous at aperture = aperture0 and aperture =
+/// aperture_cutoff.
+inline double logPenaltyDerivative(double const aperture0,
+                                   double const aperture,
+                                   double const aperture_cutoff)
+{
+    if (aperture >= aperture0)
```
Comment (wenqing) [inline MaterialLib/FractureModels/LogPenalty.h]: Is it good to compare two real numbers with  ``>=``?

File: MaterialLib/FractureModels/LogPenalty.h [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+               2 * penalty / aperture_cutoff *
+                   (2 * aperture - aperture_cutoff - aperture0);
+    }
+};
+
+inline double logPenalty(double const aperture0,
+                         double const aperture,
+                         double const aperture_cutoff)
+{
+    if (aperture >= aperture0)
```
Comment (wenqing) [inline MaterialLib/FractureModels/LogPenalty.h]: the same here.

File: MaterialLib/FractureModels/LogPenalty.h (hunk 2 above)
Comment (endJunction) [reply MaterialLib/FractureModels/LogPenalty.h]: I don't see a problem; Comparing with `if (!(aperture < aperuture0))` yields the same assembly using the [ucomisd](http://www.felixcloutier.com/x86/UCOMISD.html) instruction.

At the boundary `aperture == aperture0` the functions result is the same, but the computation (`return 1`) is way easier.

Comment (wenqing) [reply MaterialLib/FractureModels/LogPenalty.h]: I see. Thanks for the explanation.

**Comment by endJunction** [comment]
@renchao-lu Can this PR be merged (after the tests pass, of course) or do you have some additional comments?

**Comment by renchao-lu** [comment]
Thanks for your kindly asking me suggestions. I have no further comment. Please feel free.

**Pull request merged**